src/sql/workbench/contrib/welcome/page/browser/welcomePage.ts

```diff
@@ -0,0 +1,1066 @@
+				}
+			}
+			else if (e.keyCode === 40 || e.keyCode === 39) {
+				if (e.target === dropdownLastElement) {
+					dropdownFirstElement.focus();
+				} else {
+					const moveNext = <HTMLElement>document.querySelector('.move:focus').parentElement.nextElementSibling.children[0] as HTMLElement;
+					moveNext.focus();
+				}
+			}
+		});
+	}
+
+	private createPreviewModal() {
+		const modal = document.querySelector('#preview_modal') as HTMLElement;
+		const btn = document.querySelector('#tool_tip__container--mobile') as HTMLElement;
+		const span = document.querySelector('.close_icon') as HTMLElement;
+		const previewModalHeader = document.querySelector('.preview_modal__header') as HTMLElement;
+
+		btn.addEventListener('click', function () {
+			modal.classList.toggle('show');
+		});
+
+		span.addEventListener('click', function () {
+			modal.classList.remove('show');
+		});
+
+		window.addEventListener('click', (e: MouseEvent) => {
+			if (e.target === modal && modal.classList.contains('show')) {
+				modal.classList.remove('show');
+			}
+		});
+
+		btn.addEventListener('keydown', (e: KeyboardEvent) => {
+			if (e.keyCode === 13 || e.keyCode === 32) {
+				modal.classList.toggle('show');
+				modal.setAttribute('aria-hidden', 'false');
+				previewModalHeader.focus();
+			}
+			if (e.keyCode === 27) {
+				if (modal.classList.contains('show')) {
+					modal.setAttribute('aria-hidden', 'true');
+					modal.classList.remove('show');
+				}
+			}
+		});
+
+		window.addEventListener('keydown', (e: KeyboardEvent) => {
+			const target = e.target as HTMLTextAreaElement;
+			if (!target.matches('.modal') && e.keyCode === 27) {
+				if (modal.classList.contains('show')) {
+					modal.setAttribute('aria-hidden', 'true');
+					modal.classList.remove('show');
+				}
+			}
+		});
+
+		modal.addEventListener('keydown', function (e: KeyboardEvent) {
+			const previewModalBody = document.querySelector('.preview_modal__body') as HTMLElement;
+			const previewModalHeader = document.querySelector('.preview_modal__header') as HTMLElement;
+
+			if (e.keyCode === 9) {
+				e.preventDefault();
+				if (e.target === previewModalBody) {
+					previewModalHeader.focus();
+				} else {
+					previewModalBody.focus();
+				}
+			}
+		});
+	}
+
+	private createListEntries(recents: (IRecentWorkspace | IRecentFolder)[], fileService) {
+		return recents.map(recent => {
+			let relativePath: string;
+			let fullPath: URI;
+			let windowOpenable: IWindowOpenable;
+			let mtime: Date;
+			if (isRecentFolder(recent)) {
+				windowOpenable = { folderUri: recent.folderUri };
+				relativePath = recent.label || this.labelService.getWorkspaceLabel(recent.folderUri, { verbose: true });
+				fullPath = recent.folderUri;
+			} else {
+				relativePath = recent.label || this.labelService.getWorkspaceLabel(recent.workspace, { verbose: true });
+				windowOpenable = { workspaceUri: recent.workspace.configPath };
+			}
+			return fileService.resolve(fullPath).then((value) => {
```

Comment:
If you're returning a promise, mark the function as async and have the return as the signature of the function

Reply:
complete